src/audio/base_fw.c

```diff
@@ -228,13 +235,99 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
 	return 0;
 }
 
+static uint32_t basefw_set_system_time(uint32_t param_id,
+				       bool first_block,
+				       bool last_block,
+				       uint32_t data_offset,
+				       char *data)
+{
+	if (!(first_block && last_block))
+		return IPC4_INVALID_REQUEST;
+
+	global_system_time_info.host_time.val_l = (*(struct ipc4_system_time *)(data)).val_l;
+	global_system_time_info.host_time.val_u = (*(struct ipc4_system_time *)(data)).val_u;
+	uint64_t current_dsp_time = sof_cycle_get_64();
+
+	global_system_time_info.dsp_time.val_l = (uint32_t)(current_dsp_time);
+	global_system_time_info.dsp_time.val_u = (uint32_t)(current_dsp_time >> 32);
+	return IPC4_SUCCESS;
+}
+
+static uint32_t basefw_get_system_time(uint32_t *data_offset, char *data)
+{
+	struct ipc4_system_time *system_time = (struct ipc4_system_time *)data;
+
+	system_time->val_l = global_system_time_info.host_time.val_l;
+	system_time->val_u = global_system_time_info.host_time.val_u;
+	*data_offset = sizeof(struct ipc4_system_time);
+	return IPC4_SUCCESS;
+}
+
+static uint32_t basefw_get_ext_system_time(uint32_t *data_offset, char *data)
+{
+#if CONFIG_ACE_V1X_ART_COUNTER && CONFIG_ACE_V1X_RTC_COUNTER
+	struct ipc4_ext_system_time *ext_system_time = (struct ipc4_ext_system_time *)(data);
+	struct ipc4_ext_system_time ext_system_time_data = {0};
+
+	uint64_t host_time = ((uint64_t)global_system_time_info.host_time.val_u << 32)
+				| (uint64_t)global_system_time_info.host_time.val_l;
+	uint64_t dsp_time = ((uint64_t)global_system_time_info.dsp_time.val_u << 32)
+				| (uint64_t)global_system_time_info.dsp_time.val_l;
+	(void)host_time;
+	(void)dsp_time;
+
+	uint64_t art = 0;
+	uint64_t wallclk = 0;
+	uint64_t rtc = 0;
+
+	const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(ace_art_counter));
+
+	if (!dev) {
+		LOG_DBG("board: ACE V1X device binding failed");
+		return IPC4_FAILURE;
+	}
+
+	counter_get_value_64(dev, &art);
```

Comment:
Make sure this value is provided by ART HAL function on Intel platforms.

Reply:
@mmaka1 Please correct me if I'm wrong, but I think that the whole point of what we're doing here is to abandon those HAL functions in favor of Zephyr legitimate drivers. In case of RTC and ART, two additional Intel drivers have been added to Zephyr (PR: https://github.com/zephyrproject-rtos/zephyr/pull/49175). They are supposed to do the same thing as the HAL functions that you've mentioned. To ensure that we are getting the right values, we pass the the proper device as an argument of counter_get_value_64() function.